### PR TITLE
sd-bus: close peer pidfd

### PR DESCRIPTION
**Pull request by cgzones**
Do not leak the pidfd of the peer.

Fixes: 71be64064c9e ("sd-bus: add pidfd to the sd_bus_creds structure")

### CODE REVIEW

File: src/libsystemd/sd-bus/sd-bus.c [hunk 1]
```diff
@@ -152,6 +152,12 @@ void bus_close_inotify_fd(sd_bus *b) {
         b->n_inotify_watches = 0;
 }
 
+static void bus_close_peer_pidfd(sd_bus *b) {
+        assert(b);
+
+        b->pidfd = safe_close(b->pidfd);
```
Comment (YHNdnzj) [inline src/libsystemd/sd-bus/sd-bus.c:160]: There's no need to make this a dedicated function?

File: src/libsystemd/sd-bus/sd-bus.c [hunk 2]
```diff
@@ -194,6 +200,7 @@ static sd_bus* bus_free(sd_bus *b) {
 
         bus_close_io_fds(b);
         bus_close_inotify_fd(b);
+        bus_close_peer_pidfd(b);
```
Comment (yuwata) [inline src/libsystemd/sd-bus/sd-bus.c]: Maybe, better to introduce `bus_close_fds()` or so?

**Comment by yuwata** [comment]
Note, the pidfd is actually leaked.
```diff
diff --git a/src/libsystemd/sd-bus/test-bus-peersockaddr.c b/src/libsystemd/sd-bus/test-bus-peersockaddr.c
index a7bba17a98..255ae68244 100644
--- a/src/libsystemd/sd-bus/test-bus-peersockaddr.c
+++ b/src/libsystemd/sd-bus/test-bus-peersockaddr.c
@@ -6,6 +6,7 @@
 #include "sd-bus.h"
 
 #include "bus-dump.h"
+#include "bus-internal.h"
 #include "bus-util.h"
 #include "fd-util.h"
 #include "process-util.h"
@@ -122,6 +123,8 @@ static void *server(void *p) {
                 }
         }
 
+        log_info("server: bus->pidfd=%i", bus->pidfd);
+
         return NULL;
 }
 
@@ -140,6 +143,8 @@ static void* client(void *p) {
         assert_se(sd_bus_message_read(reply, "s", &z) >= 0);
         assert_se(streq_ptr(z, "bar"));
 
+        log_info("client: bus->pidfd=%i", bus->pidfd);
+
         return NULL;
 }
 
@@ -173,6 +178,13 @@ TEST(description) {
 
         assert_se(pthread_join(s, NULL) == 0);
         assert_se(pthread_join(c, NULL) == 0);
+
+        for (int i = 0; i < 10; i++) {
+                _cleanup_free_ char *path = NULL;
+
+                if (fd_get_path(i, &path) >= 0)
+                        log_info("%i: %s", i, path);
+        }
 }
 
 DEFINE_TEST_MAIN(LOG_INFO);
```
```
$ ./test-bus-peersockaddr
/* test_description */
PID=2448239
PIDFD=yes
EUID=1000
EGID=1000
SupplementaryGIDs=4 10 135 1000
Comm=test-bus-peerso
Description=wuffwuff
server: bus->pidfd=7
client: bus->pidfd=6
0: /dev/pts/2
1: /dev/pts/2
2: /dev/pts/2
6: anon_inode:[pidfd]
7: anon_inode:[pidfd]
```

**Comment by poettering** [comment]
lgtm. i'd even merge it as is, but the rework @yuwata/@YHNdnzj asked for sounds good to.

**Comment by bluca** [comment]
@cgzones if you don't have time to fix this today, I will push changes here myself in a couple of hours or so, as this is a bugfix that should really be in the next rc

**Comment by bluca** [comment]
> @cgzones if you don't have time to fix this today, I will push changes here myself in a couple of hours or so, as this is a bugfix that should really be in the next rc

Updated as requested in the review

**Comment by poettering** [comment]
lgtm

**Review by YHNdnzj — APPROVED** [review]

**Review by yuwata — APPROVED** [review]

**Review by keszybz — APPROVED** [review]
LGTM too.

File: src/libsystemd/sd-bus/sd-bus.c [hunk 3]
```diff
@@ -152,6 +152,14 @@ void bus_close_inotify_fd(sd_bus *b) {
         b->n_inotify_watches = 0;
 }
 
+static void bus_close_fds(sd_bus *b) {
+        assert(b);
+
+        bus_close_io_fds(b);
+        bus_close_inotify_fd(b);
+        b->pidfd = safe_close(b->pidfd);
```
Comment (evverx) [inline src/libsystemd/sd-bus/sd-bus.c:160]: Valgrind isn't happy about this
```
Bus bus-api-resolve: changing state RUNNING → CLOSED
==11089== File descriptor 20: /proc/sys/kernel/ngroups_max is already closed
==11089==    at 0x557CEA4: close (in /usr/lib64/libc.so.6)
==11089==    by 0x4B45B62: close_nointr (fd-util.c:41)
==11089==    by 0x4B45BBF: safe_close (fd-util.c:75)
==11089==    by 0x4C11AE9: bus_close_fds (sd-bus.c:160)
==11089==    by 0x4C18D8A: sd_bus_close (sd-bus.c:1811)
==11089==    by 0x4C1F6A8: quit_callback (sd-bus.c:3758)
==11089==    by 0x4CB5868: source_dispatch (sd-event.c:4262)
==11089==    by 0x4CB6054: dispatch_exit (sd-event.c:4384)
==11089==    by 0x4CB7A38: sd_event_dispatch (sd-event.c:4836)
==11089==    by 0x4CB7F64: sd_event_run (sd-event.c:4904)
==11089==    by 0x4CB8148: sd_event_loop (sd-event.c:4926)
==11089==    by 0x477CA0: run (resolved.c:92)
==11089==  Previously closed
==11089==    at 0x55828CB: __close_nocancel (in /usr/lib64/libc.so.6)
==11089==    by 0x55777F1: sysconf (in /usr/lib64/libc.so.6)
==11089==    by 0x4B9D1B4: getpeergroups (socket-util.c:934)
==11089==    by 0x4C0B244: bus_get_peercred (bus-socket.c:676)
==11089==    by 0x4C0B778: bus_socket_start_auth (bus-socket.c:740)
==11089==    by 0x4C0CD61: bus_socket_connect (bus-socket.c:1121)
==11089==    by 0x4C16559: bus_start_address (sd-bus.c:1149)
==11089==    by 0x4C16BF3: sd_bus_start (sd-bus.c:1243)
==11089==    by 0x4935701: bus_open_system_watch_bind_with_description (bus-util.c:713)
==11089==    by 0x414412: manager_connect_bus (resolved-bus.c:2262)
==11089==    by 0x462887: manager_new (resolved-manager.c:717)
==11089==    by 0x477A29: run (resolved.c:72)
==11089==  Originally opened
==11089==    at 0x5582B72: __open_nocancel (in /usr/lib64/libc.so.6)
==11089==    by 0x55778DA: sysconf (in /usr/lib64/libc.so.6)
==11089==    by 0x4B9D1B4: getpeergroups (socket-util.c:934)
==11089==    by 0x4C0B244: bus_get_peercred (bus-socket.c:676)
==11089==    by 0x4C0B778: bus_socket_start_auth (bus-socket.c:740)
==11089==    by 0x4C0CD61: bus_socket_connect (bus-socket.c:1121)
==11089==    by 0x4C16559: bus_start_address (sd-bus.c:1149)
==11089==    by 0x4C16BF3: sd_bus_start (sd-bus.c:1243)
==11089==    by 0x4935701: bus_open_system_watch_bind_with_description (bus-util.c:713)
==11089==    by 0x414412: manager_connect_bus (resolved-bus.c:2262)
==11089==    by 0x462887: manager_new (resolved-manager.c:717)
==11089==    by 0x477A29: run (resolved.c:72)
```

Comment (yuwata) [reply src/libsystemd/sd-bus/sd-bus.c:160]: @evverx Feel free to open a new issue like that. Otherwise, we may easily forget or overlooked your comment.

Comment (yuwata) [reply src/libsystemd/sd-bus/sd-bus.c:160]: Filed as #32995.

**Pull request merged**